index.js

```diff
@@ -780,22 +734,129 @@
+  function Function_(types) {
+    function format(outer, inner) {
+      var xs = types.map(function(t, idx) {
+        return unless(t.type === RECORD || isEmpty(t.keys),
+                      stripOutermostParens,
+                      inner('$' + String(idx + 1))(String(t)));
+      });
+      var parenthesize = wrap(outer('('))(outer(')'));
+      return parenthesize(unless(types.length === 2,
+                                 parenthesize,
+                                 init(xs).join(outer(', '))) +
+                          outer(' -> ') +
+                          last(xs));
+    }
+
+    var $keys = [];
+    var $types = {};
+    types.forEach(function(t, idx) {
+      var k = '$' + String(idx + 1);
+      $keys.push(k);
+      $types[k] = {extractor: K([]), type: t};
+    });
+
+    return new _Type(FUNCTION,
+                     '',
+                     '',
+                     format,
+                     AnyFunction,
+                     K(true),
+                     $keys,
+                     $types);
+  }
+
+  //# NonEmpty :: Type -> Type
+  //.
+  //. Constructor for non-empty types. `$.NonEmpty($.String)`, for example, is
+  //. the type comprising every [`String`][] value except `''`.
+  //.
+  //. The given type must satisfy the [Monoid][] and [Setoid][] specifications.
+  var NonEmpty = UnaryType(
+    'sanctuary-def/NonEmpty',
+    functionUrl('NonEmpty'),
+    Any,
+    function(x) {
+      return Z.Monoid.test(x) &&
+             Z.Setoid.test(x) &&
+             !Z.equals(x, Z.empty(x.constructor));
+    },
+    function(monoid) { return [monoid]; }
   );
 
-  //# ValidNumber :: Type
+  //# Nullable :: Type -> Type
   //.
-  //. Type comprising every [`Number`][] value except `NaN`.
-  var ValidNumber = NullaryTypeWithUrl(
-    'sanctuary-def/ValidNumber',
-    function(x) { return Number_._test(x) && !isNaN(x); }
+  //. Constructor for types which include `null` as a member.
+  var Nullable = UnaryTypeWithUrl(
+    'sanctuary-def/Nullable',
+    Any,
+    K(true),
+    function(nullable) {
+      // eslint-disable-next-line eqeqeq
+      return nullable === null ? [] : [nullable];
+    }
+  );
+
+  //# Pair :: Type -> Type -> Type
+  //.
+  //. Constructor for tuple types of length 2. Arrays are said to represent
+  //. tuples. `['foo', 42]` is a member of `Pair String Number`.
+  var Pair = BinaryTypeWithUrl(
+    'sanctuary-def/Pair',
+    Array_(Any),
+    function(x) { return x.length === 2; },
+    function(pair) { return [pair[0]]; },
+    function(pair) { return [pair[1]]; }
+  );
+
+  //# Strict :: Type -> Type
+  //.
+  //. Constructor for strict record types. For example:
+  //. `$.Strict($.Record($.Any, {name: $.String}))`, is the type comprising
+  //. every object with a `name :: String` property, without any other
+  //. properties.
+  function Strict(t) {
+    function test(x) {
+      return Object.keys(x).length === t.keys.length;
```

Comment:
We use `keys.every(k => k in o)` as a `test` for record types. Will this lead to problems?

```js
> 'then' in new Promise(() => {})
true

> (function(o) { var keys = []; for (var k in o) keys.push(k); return keys; }(new Promise(() => {})))
[]
```

This would mean that:

```js
const Thenable = $.RecordType({then: $.AnyFunction});
const StrictThenable = $.Strict(Thenable);

Thenable.validate(new Promise(noop)) // => Right
StrictThenable.validate(new Promise(noop)) // => Left
```

I'm not sure if that's desired. Using enumerable keys over "own" keys is an improvement though, so I think I should implement your suggestion nonetheless.